File: examples/blackjack/train-pettingzoo.js
/*
 * examples/blackjack/train-pettingzoo.js
 * Multi-agent training using PettingZoo AEC interface
 *
 * Demonstrates the AEC (Agent Environment Cycle) training loop where
 * agents take turns acting sequentially.
 */
import { BlackjackAEC, ACTION_HIT, ACTION_STAND, ACTION_DOUBLE } from "./BlackjackAEC.js";

/**
 * Simple policy that uses basic blackjack strategy
 * @param {number[]} observation - [handValue, dealerUpcard, isSoft, canDouble, canSplit, canInsurance, deckRatio]
 * @param {boolean[]} actionMask - Valid actions [Hit, Stand, Double, Split, Insurance]
 * @returns {number} Action to take
 */
function basicStrategyPolicy(observation, actionMask) {
  const handValue = observation[0] * 30; // Denormalize
  const dealerUpcard = observation[1] * 11; // Denormalize
  const isSoft = observation[2] > 0.5;
  const canDouble = actionMask[ACTION_DOUBLE];

  // Simplified basic strategy
  if (isSoft) {
    // Soft hands (Ace counted as 11)
    if (handValue >= 19) return ACTION_STAND;
    if (handValue === 18) {
      if (dealerUpcard >= 9) return ACTION_HIT;
      return ACTION_STAND;
    }
    // Soft 17 or less - hit
    return ACTION_HIT;
  }

  // Hard hands
  if (handValue >= 17) return ACTION_STAND;

  if (handValue >= 12 && handValue <= 16) {
    // Stand against dealer 2-6, hit against 7+
    if (dealerUpcard >= 2 && dealerUpcard <= 6) return ACTION_STAND;
    return ACTION_HIT;
  }

  if (handValue === 11) {
    // Double down on 11 if possible
    if (canDouble) return ACTION_DOUBLE;
    return ACTION_HIT;
  }

  if (handValue === 10) {
    // Double down on 10 against dealer 2-9 if possible
    if (canDouble && dealerUpcard >= 2 && dealerUpcard <= 9) return ACTION_DOUBLE;
    return ACTION_HIT;
  }

  if (handValue === 9) {
    // Double down on 9 against dealer 3-6 if possible
    if (canDouble && dealerUpcard >= 3 && dealerUpcard <= 6) return ACTION_DOUBLE;
    return ACTION_HIT;
  }

  // Hit on 8 or less
  return ACTION_HIT;
}

/**
 * Random policy for comparison
 * @param {number[]} observation
 * @param {boolean[]} actionMask
 * @returns {number}
 */
function randomPolicy(observation, actionMask) {
  // Only consider Hit and Stand for simplicity
  const validActions = [ACTION_HIT, ACTION_STAND].filter((a) => actionMask[a]);
  if (validActions.length === 0) return ACTION_STAND;
  return validActions[Math.floor(Math.random() * validActions.length)];
}

/**
 * Run training episodes with the AEC environment
 */
async function runTraining() {
  console.log("=== PettingZoo AEC Multi-Agent Blackjack Training ===\n");

  const env = new BlackjackAEC({
    numAgents: 2,
    numDecks: 6,
    initialBankroll: 1000,
    defaultBet: 10,
    agentNames: ["BasicStrategy", "RandomAgent"],
  });

  const episodes = 100;
  const results = {
    BasicStrategy: { wins: 0, losses: 0, ties: 0, totalReward: 0 },
    RandomAgent: { wins: 0, losses: 0, ties: 0, totalReward: 0 },
  };

  // Assign policies to agents
  const policies = {
    BasicStrategy: basicStrategyPolicy,
    RandomAgent: randomPolicy,
  };

  for (let episode = 1; episode <= episodes; episode++) {
    // Reset environment
    await env.reset();

    // AEC loop - agents take turns
    let done = false;
    let steps = 0;
    const maxSteps = 50; // Safety limit

    while (!done && steps < maxSteps) {
      const agent = env.agentSelection();
      if (!agent) break;

      // Check if this agent is terminated
      const terminations = env.terminations();
      if (terminations[agent]) {
        break;
      }

      // Get observation and action mask
      const observation = env.observe(agent);
      const actionMask = env.actionMask(agent) || [true, true, false, false, false];

      // Get action from policy
      const policy = policies[agent];
      const action = policy(observation, actionMask);

      // Step environment
      await env.step(action);

      // Check if all agents are done
      const allTerminations = env.terminations();
      done = Object.values(allTerminations).every((t) => t);

      steps++;
    }

    // Collect rewards
    const rewards = env.cumulativeRewards();

    for (const agent of env.possibleAgents) {
      const reward = rewards[agent] || 0;
      results[agent].totalReward += reward;

      if (reward > 0) results[agent].wins++;
      else if (reward < 0) results[agent].losses++;
      else results[agent].ties++;
    }

    // Log progress every 10 episodes
    if (episode % 10 === 0) {
      console.log(`Episode ${episode}/${episodes} complete`);
    }
  }

  // Print final results
  console.log("\n=== Training Results ===\n");

  for (const agent of env.possibleAgents) {
    const r = results[agent];
    const avgReward = r.totalReward / episodes;
    const winRate = ((r.wins / episodes) * 100).toFixed(1);

    console.log(`${agent}:`);
    console.log(`  Wins: ${r.wins}, Losses: ${r.losses}, Ties: ${r.ties}`);
    console.log(`  Win Rate: ${winRate}%`);
    console.log(`  Total Reward: $${r.totalReward.toFixed(2)}`);
    console.log(`  Avg Reward/Episode: $${avgReward.toFixed(2)}\n`);
  }

  return results;
}

/**
 * Run a demo game with rendering
 */
async function runDemo() {
  console.log("=== Demo Game ===\n");

  const env = new BlackjackAEC({
    numAgents: 2,
    numDecks: 6,
    initialBankroll: 1000,
    defaultBet: 10,
    agentNames: ["Player1", "Player2"],
  });

  await env.reset();
  env.render();

  const policies = {
    Player1: basicStrategyPolicy,
    Player2: randomPolicy,
  };

  let done = false;
  let steps = 0;

  while (!done && steps < 20) {
    const agent = env.agentSelection();
    if (!agent) break;

    const terminations = env.terminations();
    if (terminations[agent]) break;

    const observation = env.observe(agent);
    const actionMask = env.actionMask(agent) || [true, true, false, false, false];
    const policy = policies[agent];
    const action = policy(observation, actionMask);

    const actionNames = ["Hit", "Stand", "Double", "Split", "Insurance"];
    console.log(`${agent} chooses: ${actionNames[action]}`);

    await env.step(action);
    env.render();

    done = Object.values(env.terminations()).every((t) => t);
    steps++;
  }

  // Show final results
  const rewards = env.cumulativeRewards();
  console.log("=== Final Rewards ===");
  for (const [agent, reward] of Object.entries(rewards)) {
    console.log(`${agent}: $${reward}`);
  }
}

// Main execution
console.log("Starting PettingZoo AEC Training...\n");

runDemo()
  .then(() => {
    console.log("\n--- Starting Full Training ---\n");
    return runTraining();
  })
  .then(() => {
    console.log("Training complete!");
  })
  .catch((err) => {
    console.error("Error:", err);
  });

File: hypertoken-rl/python/examples/train_pettingzoo.py
#!/usr/bin/env python3
"""
Train multiple agents using PettingZoo AEC API.

This example demonstrates the standard PettingZoo AEC loop for
multi-agent environments where agents take turns acting.

Prerequisites:
    pip install hypertoken

Usage:
    # Start the server first (with 3 agents):
    npx tsx bridge/server.ts --env blackjack --agents 3 --port 9999

    # Then run:
    python train_pettingzoo.py
"""

import numpy as np
from hypertoken import HyperTokenAECEnv


def random_policy(observation, action_space, action_mask=None):
    """Random policy that respects action mask."""
    if action_mask is not None:
        # Sample from valid actions only
        valid_actions = np.where(action_mask)[0]
        if len(valid_actions) > 0:
            return np.random.choice(valid_actions)
    return action_space.sample()


def basic_strategy_policy(observation, action_space, action_mask=None):
    """
    Basic blackjack strategy based on hand value.

    Observation format:
        [handValue, dealerUpcard, isSoft, canDouble, canSplit, canInsurance, deckRatio]
        All values normalized to 0-1 range.
    """
    # Denormalize hand value: 0-1 maps to roughly 0-30
    hand_value = int(observation[0] * 30)

    # Denormalize dealer upcard: 0-1 maps to 0-11
    dealer_upcard = int(observation[1] * 11)

    is_soft = observation[2] > 0.5
    can_double = observation[3] > 0.5
    can_split = observation[4] > 0.5

    # Basic strategy
    action = 1  # Default: stand

    if is_soft:
        # Soft hand strategy
        if hand_value < 18:
            action = 0  # Hit
        elif hand_value == 18 and dealer_upcard in [9, 10, 11]:
            action = 0  # Hit
    else:
        # Hard hand strategy
        if hand_value < 12:
            action = 0  # Hit
        elif hand_value < 17:
            if dealer_upcard >= 7:
                action = 0  # Hit
            else:
                action = 1  # Stand
        else:
            action = 1  # Stand

    # Double down opportunity
    if can_double and hand_value in [10, 11]:
        if dealer_upcard < 10:
            action = 2  # Double

    # Validate against mask
    if action_mask is not None:
        if not action_mask[action]:
            # Fall back to hit or stand
            if action_mask[0]:
                action = 0
            elif action_mask[1]:
                action = 1
            else:
                # Pick any valid action
                valid = np.where(action_mask)[0]
                action = valid[0] if len(valid) > 0 else 1

    return action


def run_episode(env, policies):
    """
    Run a single episode with the given policies.

    Args:
        env: HyperTokenAECEnv instance
        policies: Dict mapping agent names to policy functions

    Returns:
        Dict of final rewards per agent
    """
    env.reset()

    for agent in env.agent_iter():
        obs, reward, terminated, truncated, info = env.last()

        if terminated or truncated:
            action = None
        else:
            action_mask = env.action_mask(agent)
            policy = policies.get(agent, random_policy)
            action = policy(obs, env.action_space(agent), action_mask)

        env.step(action)

    return env.rewards()


def main():
    """Run multi-agent training example."""
    print("=" * 60)
    print("HyperToken Blackjack - PettingZoo Multi-Agent Training")
    print("=" * 60)

    print("\nConnecting to HyperToken environment...")

    with HyperTokenAECEnv("ws://localhost:9999") as env:
        print(f"  Connected!")
        print(f"  Agents: {env.possible_agents}")
        print(f"  Observation space: {env.observation_space(env.possible_agents[0])}")
        print(f"  Action space: {env.action_space(env.possible_agents[0])}")

        # Define policies for each agent
        # Agent 0 uses basic strategy, others use random
        policies = {}
        for i, agent in enumerate(env.possible_agents):
            if i == 0:
                policies[agent] = basic_strategy_policy
            else:
                policies[agent] = random_policy

        print(f"\nPolicies:")
        for agent, policy in policies.items():
            print(f"  {agent}: {policy.__name__}")

        # Run episodes
        num_episodes = 100
        agent_rewards = {agent: [] for agent in env.possible_agents}

        print(f"\nRunning {num_episodes} episodes...")
        for episode in range(num_episodes):
            rewards = run_episode(env, policies)

            for agent, reward in rewards.items():
                agent_rewards[agent].append(reward)

            if (episode + 1) % 20 == 0:
                print(f"  Episode {episode + 1}/{num_episodes}")

        # Print results
        print("\n" + "=" * 60)
        print("Results")
        print("=" * 60)

        for agent in env.possible_agents:
            rewards = agent_rewards[agent]
            avg_reward = np.mean(rewards)
            win_rate = np.mean([1 if r > 0 else 0 for r in rewards]) * 100
            policy_name = policies[agent].__name__

            print(f"\n{agent} ({policy_name}):")
            print(f"  Average reward: {avg_reward:.2f}")
            print(f"  Win rate: {win_rate:.1f}%")
            print(f"  Total earnings: {sum(rewards):.2f}")

        # Compare basic strategy vs random
        print("\n" + "=" * 60)
        print("Policy Comparison")
        print("=" * 60)

        basic_agent = env.possible_agents[0]
        random_agents = env.possible_agents[1:]

        basic_avg = np.mean(agent_rewards[basic_agent])
        random_avg = np.mean([np.mean(agent_rewards[a]) for a in random_agents])

        print(f"  Basic strategy avg reward: {basic_avg:.2f}")
        print(f"  Random policy avg reward: {random_avg:.2f}")
        print(f"  Improvement: {((basic_avg - random_avg) / abs(random_avg) * 100) if random_avg != 0 else 0:.1f}%")


if __name__ == "__main__":
    main()
